### PR TITLE
Added `flake8-bugbear` to `pre-commit` hooks

### LINKED ISSUE #9369: Add `flake8-bugbear` to `pre-commit` hooks
In https://github.com/dask/distributed/pull/6809 `distributed` added `flake8-bugbear` to their `pre-commit` hooks (thanks @hendrikmakait!) We might consider doing the same here. 

Copying from https://github.com/dask/distributed/issues/6662:

> flake8-bugbear (https://github.com/PyCQA/flake8-bugbear) is a plugin for flake8 that contains additional rules aimed at finding likely bugs and design problems in the code. Notable rules include:

    B006: Do not use mutable data structures for argument defaults.
    B018: Found useless expression.

### PR DESCRIPTION
**Pull request by lukeconibear**
### Summary of changes in this PR

- [x] `.pre-commit-config.yaml`: added `flake8-bugbear` hook for `pre-commit`
- [x] `setup.cfg`: excluded `B011` for tests following https://github.com/dask/distributed/pull/6809
- [x] Misc. warnings:
  - [x] `dask/array/creation.py`: replaced assert False with raise error
    - _**B011**: Do not call assert False since python -O removes these calls. Instead callers should raise `AssertionError()`._
  - [x] `dask/array/tests/test_dispatch.py`: replaced `setattr` for constant attribute
    - _**B010**: Do not call `setattr(x, 'attr', val)`, instead use normal property access: `x.attr = val`. There is no additional safety in using `setattr` if you know the attribute name ahead of time._ 

### Depends on these PRs merging first

- [x] Individual PRs per warning from `flake8-bugbear`:
  - [x] #9458
  - [x] #9459
  - [x] #9460
  - [x] #9461   
  - [x] #9462 

### General

- [x] Closes #9369 
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

**Comment by ncclementi** [comment]
add to allowlist

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -32,6 +32,8 @@ repos:
       hooks:
       - id: flake8
         language_version: python3
+        additional_dependencies:
```
Comment (jrbourbeau) [inline .pre-commit-config.yaml:35]: Could you add a note to the note top of this file similar to the one over in `dask/distributed`? 

https://github.com/dask/distributed/blob/1314ebb8671493290d284bf24b1ed1da1749f90c/.pre-commit-config.yaml#L1-L2

Comment (lukeconibear) [reply .pre-commit-config.yaml:37]: Sure, added

**Review by jrbourbeau — APPROVED** [review]
Thanks for working on this @lukeconibear!

**Comment by lukeconibear** [comment]
Great, thanks very much for all your help!

**Pull request merged**